### PR TITLE
Stats - pre-review - Refactoring of Stats Service and Fragments

### LINKED ISSUE #2561: Refactor Stats fragments by making them highly modulars and less "Activity" dependent
A placeholder for this week's activity.

### PR DESCRIPTION
**Pull request by daniloercoli**
May I get a code pre-review of this upcoming PR that basically fixes #2561, and gives us the capability of introduce single stats fragments everywhere in the app?

- All of the network interactions are now handled in the Stats Service.
- Stats service was simplified a lot and reduced to minimum terms. Don't know why we had that complex service till now.
- The Stats service can now handle multiple blogs, and stats at the same time. This gives us the ability of composing different type of stats screen. EX: an overview stats screen for all of the blogs of the users.
- Each Stats fragment basically calls the service, and ask to load the data for a particular blog_id/stats_type and other optional parameters.
- StatsActivity (the main activity) now correctly handles the Jetpack installation, and also checks if the network is available and shows/dismisses the loading indicator correctly.
- StatsActivity was cleaned a bit, and removed all of that code that handled concurrency of fragment refresh.
- StatsViewAll was simplified, and removed much of the redundant code

### CODE REVIEW

**Comment by tonyr59h** [comment]
> May I get a code pre-review

Does this mean no merge?

**Comment by daniloercoli** [comment]
Yes, no merge for now. I will submit a "real" PR early next week. All depends if the direction I'm going is OK,  or at least better than the previous code.

**Comment by tonyr59h** [comment]
Switching the timeframe (days/weeks/etc...) before the selected one finishes loading results in [no data](https://cldup.com/ftLGFp9son.png) in many of the stats modules. Maybe disable the spinner while new data is being loaded.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/service/StatsService.java [hunk 1]
```diff
@@ -136,108 +135,124 @@ private void stopRefresh() {
...
+        EventBus.getDefault().post(new StatsEvents.UpdateStatusChanged(true));
 
-                AppLog.i(T.STATS, "Update started for blogID - " + blogId + " with the following period: " + period
-                        + " on the following date: " + mServiceRequestedDate);
+        RestListener vListener = new RestListener(sectionToUpdate, blogId, timeframe, date);
+        String path = String.format("/sites/%s/stats/", blogId);
+        synchronized (mStatsNetworkRequests) {
+            switch (sectionToUpdate) {
+                case VISITS:
+                    path = String.format(path + "visits?unit=%s&quantity=10&date=%s", period, date);
```
Comment (tonyr59h) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/service/StatsService.java]: Might be a good idea to throw these strings into constants.

**Comment by tonyr59h** [comment]
Definitely moving in the right direction. The StatsActivity and StatsService simplifications are welcome.

> and ask to load the data for a particular blog_id/stats_type

So if we wanted to we could have individual stats fragments querying different blogs?

**Comment by daniloercoli** [comment]
<blockquote>Switching the timeframe (days/weeks/etc...) before the selected one finishes loading results in no data in many of the stats modules. Maybe disable the spinner while new data is being loaded.</blockquote>
We can disable the spinner, but IIRC we've decided to mimic the web experience where you can switch between timeframe even if the previous loading phase is not done yet. 
Note, that what you see during the loaidng phase is the classic "empty module" (gray box where the content will be), that is only shown during the loading phase. 
The "no-data" version is completely different and contains a string that explains there is no data....

**Comment by daniloercoli** [comment]
<blockquote>So if we wanted to we could have individual stats fragments querying different blogs?</blockquote>
Not with the current version of the client/server code. One single Stats android module/fragment can map 1 or more REST Endpoint. Since each REST endpoint is tied to a specific blog, the only option you have is to create an activity with two different fragments in it, that query the same endpoint with different parameters.

Anyway, the ability of loading Stats for different blog "inside" the same fragment/module is never emerged during the analysis, at least for now. If we want to address this necessity, I think, we should also modify the rest endpoints and return data for all of the user's blogs, otherwise to compose a single stats module we have to make a lot of REST calls.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/service/StatsService.java (hunk 1 above)
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/service/StatsService.java]: I've cleaned the code a bit. Will run the app during the weekend and open a PR on monday if everything goes fine on my tests.

**Pull request closed without merging**